File: aspnet-in-practice/06/VB/CustomControls.Web/CustomControls.Composite/PostControl.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Web.UI.WebControls
Imports System.Web.UI

Public Class PostControl
	Inherits WebControl
	Implements IPostBackEventHandler
	Public Property Value() As DateTime
		Get
			Return CType(If(ViewState("Value"), DateTime.MinValue), DateTime)
		End Get
		Set(ByVal value As DateTime)
			ViewState("Value") = value
		End Set
	End Property

	Protected Overrides Sub RenderContents(ByVal writer As HtmlTextWriter)
		Dim postBackLink As String = Page.ClientScript.GetPostBackClientHyperlink(Me, Value.ToString(), True)
		Dim link As New HyperLink()
		link.NavigateUrl = postBackLink
		link.Text = "Test PostBack"
		link.RenderControl(writer)
	End Sub

	Public Sub RaisePostBackEvent(ByVal eventArgument As String) Implements IPostBackEventHandler.RaisePostBackEvent
		' get the value from post back - your multiple postback params logic here
		Value = DateTime.Parse(eventArgument)

		' fire the event
		OnValueChanged(EventArgs.Empty)
	End Sub

	Public Event ValueChanged As EventHandler

	Protected Sub OnValueChanged(ByVal e As EventArgs)
		RaiseEvent ValueChanged(Me, e)
	End Sub

End Class

File: aspnet-in-practice/06/CS/CustomControls.Composite/PostControl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web.UI.WebControls;
using System.Web.UI;

namespace CustomControls.Composite
{
	public class PostControl : WebControl, IPostBackEventHandler
	{
		public DateTime Value
		{
			get
			{
				return (DateTime)(ViewState["Value"] ?? DateTime.MinValue);
			}
			set
			{
				ViewState["Value"] = value;
			}
		}

		protected override void RenderContents(HtmlTextWriter writer)
		{
			string postBackLink = Page.ClientScript.GetPostBackClientHyperlink(this, Value.ToString(), true);
			HyperLink link = new HyperLink();
			link.NavigateUrl = postBackLink;
			link.Text = "Test PostBack";
			link.RenderControl(writer);
		}

		public void RaisePostBackEvent(string eventArgument)
		{
			// get the value from post back - your multiple postback params logic here
			Value = DateTime.Parse(eventArgument);

			// fire the event
			OnValueChanged(EventArgs.Empty);
		}

		public event EventHandler ValueChanged;

		protected void OnValueChanged(EventArgs e)
		{
			if (ValueChanged != null)
				ValueChanged(this, e);
		}

	}
}